<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial transcriptomic clustering • Banksy</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Spatial transcriptomic clustering">
<meta property="og:description" content="Banksy is an R package that incorporates spatial information to 
    cluster cells in a feature space (e.g. gene expression). To incorporate 
    spatial information, BANKSY computes the mean neighborhood expression and 
    azimuthal Gabor filters that capture gene expression gradients. These 
    features are combined with the cell's own expression to embed cells in a 
    neighbor-augmented product space which can then be clustered, allowing for
    accurate and spatially-aware cell typing and tissue domain segmentation. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Banksy</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/parameter-selection.html">BANKSY parameter selection (VeraFISH Mouse Hippocampus)</a>
    </li>
    <li>
      <a href="articles/domain-segment.html">Domain segmentation (STARmap PLUS Mouse brain)</a>
    </li>
    <li>
      <a href="articles/batch-correction.html">Spatial data integration with Harmony (10x Visium Human DLPFC)</a>
    </li>
    <li>
      <a href="articles/multi-sample.html">Analysing multiple datasets (10x Visium Human DLPFC)</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/prabhakarlab/Banksy/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>BANKSY is a method for clustering spatial omics data by augmenting the features of each cell with both an average of the features of its spatial neighbors along with neighborhood feature gradients. By incorporating neighborhood information for clustering, BANKSY is able to</p>
<ul>
<li>improve cell-type assignment in noisy data</li>
<li>distinguish subtly different cell-types stratified by microenvironment</li>
<li>identify spatial domains sharing the same microenvironment</li>
</ul>
<p>BANKSY is applicable to a wide array of spatial technologies (e.g. 10x Visium, Slide-seq, MERFISH, CosMX, CODEX) and scales well to large datasets. For more details, check out:</p>
<ul>
<li>the <a href="https://www.nature.com/articles/s41588-024-01664-3" class="external-link">paper</a>,</li>
<li>the <a href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41588-024-01664-3/MediaObjects/41588_2024_1664_MOESM3_ESM.pdf" class="external-link">peer review file</a>,</li>
<li>a <a href="https://x.com/shyam_lab/status/1762648072360792479?s=20" class="external-link">tweetorial</a> on BANKSY,</li>
<li>a set of <a href="https://prabhakarlab.github.io/Banksy">vignettes</a> showing basic usage,</li>
<li>usage compatibility with Seurat (<a href="https://github.com/satijalab/seurat-wrappers/blob/master/docs/banksy.md" class="external-link">here</a> and <a href="https://satijalab.org/seurat/articles/visiumhd_analysis_vignette#identifying-spatially-defined-tissue-domains" class="external-link">here</a>),</li>
<li>a <a href="https://github.com/prabhakarlab/Banksy_py" class="external-link">Python version</a> of this package,</li>
<li>a <a href="https://zenodo.org/records/10258795" class="external-link">Zenodo archive</a> containing scripts to reproduce the analyses in the paper, and the corresponding <a href="https://github.com/jleechung/banksy-zenodo" class="external-link">GitHub Pages</a> (and <a href="https://github.com/prabhakarlab/Banksy_py/tree/Banksy_manuscript" class="external-link">here</a> for analyses done in Python).</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The <em>Banksy</em> package can be installed via Bioconductor. This currently requires R <code>&gt;= 4.4.0</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'Banksy'</span><span class="op">)</span></span></code></pre></div>
<p>To install directly from GitHub instead, use</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"prabhakarlab/Banksy"</span><span class="op">)</span></span></code></pre></div>
<p>To use the legacy version of <em>Banksy</em> utilising the <code>BanksyObject</code> class, use</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"prabhakarlab/Banksy@legacy"</span><span class="op">)</span></span></code></pre></div>
<p><em>Banksy</em> is also interoperable with <a href="https://satijalab.org/seurat/" class="external-link"><em>Seurat</em></a> via <a href="https://github.com/satijalab/seurat-wrappers" class="external-link"><em>SeuratWrappers</em></a>. Documentation on how to run BANKSY on Seurat objects can be found <a href="https://github.com/satijalab/seurat-wrappers/blob/master/docs/banksy.md" class="external-link">here</a>. For installation of <em>SeuratWrappers</em> with BANKSY version <code>&gt;= 0.1.6</code>, run</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'satijalab/seurat-wrappers'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>Load <em>BANKSY</em>. We’ll also load <em>SpatialExperiment</em> and <em>SummarizedExperiment</em> for containing and manipulating the data, <em>scuttle</em> for normalization and quality control, and <em>scater</em>, <em>ggplot2</em> and <em>cowplot</em> for visualisation.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/prabhakarlab/Banksy" class="external-link">Banksy</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>Here, we’ll run <em>BANKSY</em> on mouse hippocampus data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hippocampus</span><span class="op">)</span></span>
<span><span class="va">gcm</span> <span class="op">&lt;-</span> <span class="va">hippocampus</span><span class="op">$</span><span class="va">expression</span></span>
<span><span class="va">locs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">hippocampus</span><span class="op">$</span><span class="va">locations</span><span class="op">)</span></span></code></pre></div>
<p>Initialize a SpatialExperiment object and perform basic quality control and normalization.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment.html" class="external-link">SpatialExperiment</a></span><span class="op">(</span>assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">gcm</span><span class="op">)</span>, spatialCoords <span class="op">=</span> <span class="va">locs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># QC based on total counts</span></span>
<span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">thres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">qcstats</span><span class="op">$</span><span class="va">total</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.98</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">qcstats</span><span class="op">$</span><span class="va">total</span> <span class="op">&gt;</span> <span class="va">thres</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">qcstats</span><span class="op">$</span><span class="va">total</span> <span class="op">&lt;</span> <span class="va">thres</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="va">keep</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Normalization to mean library size</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html" class="external-link">computeLibraryFactors</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">aname</span> <span class="op">&lt;-</span> <span class="st">"normcounts"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="va">aname</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/normalizeCounts.html" class="external-link">normalizeCounts</a></span><span class="op">(</span><span class="va">se</span>, log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Compute the neighborhood matrices for <em>BANKSY</em>. Setting <code>compute_agf=TRUE</code> computes both the weighted neighborhood mean (<span class="math inline">ℳ</span>) and the azimuthal Gabor filter (<span class="math inline">𝒢</span>). The number of spatial neighbors used to compute <span class="math inline">ℳ</span> and <span class="math inline">𝒢</span> are <code>k_geom[1]=15</code> and <code>k_geom[2]=30</code> respectively. We run <em>BANKSY</em> at <code>lambda=0</code> corresponding to non-spatial clustering, and <code>lambda=0.2</code> corresponding to <em>BANKSY</em> for cell-typing.</p>
<blockquote>
<p><strong>An important note about choosing the <code>lambda</code> parameter for the older <a href="https://doi.org/10.1038/s41593-020-00787-0" class="external-link">Visium v1 / v2 55um datasets</a> or the original <a href="https://doi.org/10.1038/s41596-018-0045-2" class="external-link">ST 100um technology</a>:</strong></p>
<p>For most modern high resolution technologies like Xenium, Visium HD, StereoSeq, MERFISH, STARmap PLUS, SeqFISH+, SlideSeq v2, and CosMx (and others), we recommend the usual defults for <code>lambda</code>: For cell typing, use <code>lambda = 0.2</code> (as shown below, or in <a href="https://prabhakarlab.github.io/Banksy/articles/parameter-selection.html">this vignette</a>) and for <a href="https://prabhakarlab.github.io/Banksy/articles/domain-segment.html">domain segmentation</a>, use <code>lambda = 0.8</code>. These technologies are either imaging based, having true single-cell resolution (e.g., MERFISH), or are sequencing based, having barcoded spots on the scale of single-cells (e.g., <a href="https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression" class="external-link">Visium HD</a>). We find that the usual defaults work well at this measurement resolution.</p>
<p>However, for the older <strong>Visium v1/v2</strong> or <strong>ST</strong> technologies, with their much lower resolution spots (55um and 100um diameter, respectively), we find that <code>lambda = 0.2</code> seems to work best for domain segmentation. This could be because each spot already measures the average transcriptome of several cells in a neighbourhood. It seems that <code>lambda = 0.2</code> shares enough information between these neighbourhoods to lead to good domain segmentation performance. For example, in the <a href="https://prabhakarlab.github.io/Banksy/articles/multi-sample.html">human DLPFC vignette</a>, we use <code>lambda = 0.2</code> on a Visium v1/v2 dataset. Also note that in these lower resolution technologies, each spot can have multiple cells of different types, and as such <em>cell-typing</em> is not defined for them.</p>
</blockquote>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lambda</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">k_geom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">Banksy</span><span class="fu">::</span><span class="fu"><a href="reference/computeBanksy.html">computeBanksy</a></span><span class="op">(</span><span class="va">se</span>, assay_name <span class="op">=</span> <span class="va">aname</span>, compute_agf <span class="op">=</span> <span class="cn">TRUE</span>, k_geom <span class="op">=</span> <span class="va">k_geom</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing neighbors...</span></span>
<span><span class="co">#&gt; Spatial mode is kNN_median</span></span>
<span><span class="co">#&gt; Parameters: k_geom=15</span></span>
<span><span class="co">#&gt; Done</span></span>
<span><span class="co">#&gt; Computing neighbors...</span></span>
<span><span class="co">#&gt; Spatial mode is kNN_median</span></span>
<span><span class="co">#&gt; Parameters: k_geom=30</span></span>
<span><span class="co">#&gt; Done</span></span>
<span><span class="co">#&gt; Computing harmonic m = 0</span></span>
<span><span class="co">#&gt; Using 15 neighbors</span></span>
<span><span class="co">#&gt; Done</span></span>
<span><span class="co">#&gt; Computing harmonic m = 1</span></span>
<span><span class="co">#&gt; Using 30 neighbors</span></span>
<span><span class="co">#&gt; Centering</span></span>
<span><span class="co">#&gt; Done</span></span></code></pre></div>
<p>Next, run PCA on the BANKSY matrix and perform clustering. Setting <code>use_agf=TRUE</code> uses both <span class="math inline">ℳ</span> and <span class="math inline">𝒢</span> to construct the BANKSY matrix.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">Banksy</span><span class="fu">::</span><span class="fu"><a href="reference/runBanksyPCA.html">runBanksyPCA</a></span><span class="op">(</span><span class="va">se</span>, use_agf <span class="op">=</span> <span class="cn">TRUE</span>, lambda <span class="op">=</span> <span class="va">lambda</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">Banksy</span><span class="fu">::</span><span class="fu"><a href="reference/runBanksyUMAP.html">runBanksyUMAP</a></span><span class="op">(</span><span class="va">se</span>, use_agf <span class="op">=</span> <span class="cn">TRUE</span>, lambda <span class="op">=</span> <span class="va">lambda</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">Banksy</span><span class="fu">::</span><span class="fu"><a href="reference/clusterBanksy.html">clusterBanksy</a></span><span class="op">(</span><span class="va">se</span>, use_agf <span class="op">=</span> <span class="cn">TRUE</span>, lambda <span class="op">=</span> <span class="va">lambda</span>, resolution <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span></code></pre></div>
<p>Different clustering runs can be relabeled to minimise their differences with <code>connectClusters</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">Banksy</span><span class="fu">::</span><span class="fu"><a href="reference/connectClusters.html">connectClusters</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; clust_M1_lam0.2_k50_res1.2 --&gt; clust_M1_lam0_k50_res1.2</span></span></code></pre></div>
<p>Visualise the clustering output for non-spatial clustering (<code>lambda=0</code>) and BANKSY clustering (<code>lambda=0.2</code>).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cnames</span> <span class="op">&lt;-</span> <span class="va">cnames</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^clust"</span>, <span class="va">cnames</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_nsp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"sdimx"</span>, y <span class="op">=</span> <span class="st">"sdimy"</span>,</span>
<span>    point_size <span class="op">=</span> <span class="fl">0.6</span>, colour_by <span class="op">=</span> <span class="va">cnames</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot_bank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"sdimx"</span>, y <span class="op">=</span> <span class="st">"sdimy"</span>,</span>
<span>    point_size <span class="op">=</span> <span class="fl">0.6</span>, colour_by <span class="op">=</span> <span class="va">cnames</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">plot_nsp</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">plot_bank</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-13-1.png" width="100%"></p>
<p>For clarity, we can visualise each of the clusters separately:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="va">plot_nsp</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">colour_by</span><span class="op">)</span>,</span>
<span>    <span class="va">plot_bank</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">colour_by</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-14-1.png" width="100%"></p>
<p>Visualize UMAPs of the non-spatial and BANKSY embedding:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rdnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span></span>
<span><span class="va">umap_nsp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>    dimred <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"UMAP.*lam0$"</span>, <span class="va">rdnames</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    colour_by <span class="op">=</span> <span class="va">cnames</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">umap_bank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>    dimred <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"UMAP.*lam0.2$"</span>, <span class="va">rdnames</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    colour_by <span class="op">=</span> <span class="va">cnames</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="va">umap_nsp</span>,</span>
<span>    <span class="va">umap_bank</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-15-1.png" width="100%"></p>
<details><summary>
Runtime for analysis
</summary><div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co">#&gt; Time difference of 35.28841 secs</span></span></code></pre></div>
</details><details><summary>
Session information
</summary><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.2.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Detroit</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] cowplot_1.1.3               scater_1.30.1              </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.4.4               scuttle_1.12.0             </span></span>
<span><span class="co">#&gt;  [5] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.54.1        GenomeInfoDb_1.38.6        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.36.0              S4Vectors_0.40.2           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.2.0           Banksy_0.99.12             </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0          viridisLite_0.4.2        </span></span>
<span><span class="co">#&gt;  [3] farver_2.1.1              dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [5] vipor_0.4.7               viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [7] bitops_1.0-7              fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [9] RCurl_1.98-1.14           digest_0.6.34            </span></span>
<span><span class="co">#&gt; [11] rsvd_1.0.5                lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt; [13] magrittr_2.0.3            dbscan_1.1-12            </span></span>
<span><span class="co">#&gt; [15] compiler_4.3.2            rlang_1.1.3              </span></span>
<span><span class="co">#&gt; [17] tools_4.3.2               igraph_2.0.1.1           </span></span>
<span><span class="co">#&gt; [19] utf8_1.2.4                yaml_2.3.8               </span></span>
<span><span class="co">#&gt; [21] data.table_1.15.0         knitr_1.45               </span></span>
<span><span class="co">#&gt; [23] labeling_0.4.3            S4Arrays_1.2.0           </span></span>
<span><span class="co">#&gt; [25] mclust_6.0.1              DelayedArray_0.28.0      </span></span>
<span><span class="co">#&gt; [27] abind_1.4-5               BiocParallel_1.36.0      </span></span>
<span><span class="co">#&gt; [29] withr_3.0.0               grid_4.3.2               </span></span>
<span><span class="co">#&gt; [31] fansi_1.0.6               beachmat_2.18.0          </span></span>
<span><span class="co">#&gt; [33] colorspace_2.1-0          aricode_1.0.3            </span></span>
<span><span class="co">#&gt; [35] scales_1.3.0              cli_3.6.2                </span></span>
<span><span class="co">#&gt; [37] rmarkdown_2.25            crayon_1.5.2             </span></span>
<span><span class="co">#&gt; [39] leidenAlg_1.1.2           generics_0.1.3           </span></span>
<span><span class="co">#&gt; [41] rstudioapi_0.15.0         rjson_0.2.21             </span></span>
<span><span class="co">#&gt; [43] DelayedMatrixStats_1.24.0 ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt; [45] RcppHungarian_0.3         zlibbioc_1.48.0          </span></span>
<span><span class="co">#&gt; [47] parallel_4.3.2            XVector_0.42.0           </span></span>
<span><span class="co">#&gt; [49] vctrs_0.6.5               Matrix_1.6-5             </span></span>
<span><span class="co">#&gt; [51] BiocSingular_1.18.0       BiocNeighbors_1.20.2     </span></span>
<span><span class="co">#&gt; [53] ggrepel_0.9.5             irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt; [55] beeswarm_0.4.0            magick_2.8.2             </span></span>
<span><span class="co">#&gt; [57] glue_1.7.0                codetools_0.2-19         </span></span>
<span><span class="co">#&gt; [59] uwot_0.1.16               RcppAnnoy_0.0.22         </span></span>
<span><span class="co">#&gt; [61] gtable_0.3.4              ScaledMatrix_1.10.0      </span></span>
<span><span class="co">#&gt; [63] munsell_0.5.0             tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [65] pillar_1.9.0              htmltools_0.5.7          </span></span>
<span><span class="co">#&gt; [67] GenomeInfoDbData_1.2.11   R6_2.5.1                 </span></span>
<span><span class="co">#&gt; [69] sparseMatrixStats_1.14.0  evaluate_0.23            </span></span>
<span><span class="co">#&gt; [71] sccore_1.0.4              lattice_0.22-5           </span></span>
<span><span class="co">#&gt; [73] highr_0.10                Rcpp_1.0.12              </span></span>
<span><span class="co">#&gt; [75] gridExtra_2.3             SparseArray_1.2.4        </span></span>
<span><span class="co">#&gt; [77] xfun_0.42                 pkgconfig_2.0.3</span></span></code></pre></div>
</details>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/Banksy" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/prabhakarlab/Banksy/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/prabhakarlab/Banksy/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Banksy</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Vipul Singhal <br><small class="roles"> Author </small>  </li>
<li>Joseph Lee <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-4983-4714" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Vipul Singhal, Joseph Lee.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
