<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Parameter selection (VeraFISH Mouse Hippocampus) â€¢ Banksy</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Parameter selection (VeraFISH Mouse Hippocampus)">
<meta property="og:description" content="Banksy">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Banksy</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/parameter-selection.html">BANKSY parameter selection (VeraFISH Mouse Hippocampus)</a>
    </li>
    <li>
      <a href="../articles/domain-segment.html">Domain segmentation (STARmap PLUS Mouse brain)</a>
    </li>
    <li>
      <a href="../articles/batch-correction.html">Spatial data integration with Harmony (10x Visium Human DLPFC)</a>
    </li>
    <li>
      <a href="../articles/multi-sample.html">Analysing multiple datasets (10x Visium Human DLPFC)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/prabhakarlab/Banksy/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Parameter selection (VeraFISH Mouse
Hippocampus)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/prabhakarlab/Banksy/blob/HEAD/vignettes/parameter-selection.Rmd" class="external-link"><code>vignettes/parameter-selection.Rmd</code></a></small>
      <div class="hidden name"><code>parameter-selection.Rmd</code></div>

    </div>

    
    
<p>Here, we demonstrate a grid search of clustering parameters with a
mouse hippocampus VeraFISH dataset. <em>BANKSY</em> currently provides
four algorithms for clustering the BANKSY matrix with
<em>clusterBanksy</em>: Leiden (default), Louvain, k-means, and
model-based clustering. In this vignette, we run only Leiden clustering.
See <code><a href="../reference/clusterBanksy.html">?clusterBanksy</a></code> for more details on the parameters for
different clustering methods.</p>
<div class="section level2">
<h2 id="loading-the-data">Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h2>
<p>The dataset comprises gene expression for 10,944 cells and 120 genes
in 2 spatial dimensions. See <code><a href="../reference/hippocampus.html">?Banksy::hippocampus</a></code> for more
details.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load libs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/prabhakarlab/Banksy" class="external-link">Banksy</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hippocampus</span><span class="op">)</span></span>
<span><span class="va">gcm</span> <span class="op">&lt;-</span> <span class="va">hippocampus</span><span class="op">$</span><span class="va">expression</span></span>
<span><span class="va">locs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">hippocampus</span><span class="op">$</span><span class="va">locations</span><span class="op">)</span></span></code></pre></div>
<p>Here, <code>gcm</code> is a gene by cell matrix, and
<code>locs</code> is a matrix specifying the coordinates of the centroid
for each cell.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gcm</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;         cell_1276 cell_8890 cell_691 cell_396 cell_9818</span></span>
<span><span class="co">#&gt; Sparcl1        45         0       11       22         0</span></span>
<span><span class="co">#&gt; Slc1a2         17         0        6        5         0</span></span>
<span><span class="co">#&gt; Map            10         0       12       16         0</span></span>
<span><span class="co">#&gt; Sqstm1         26         0        0        2         0</span></span>
<span><span class="co">#&gt; Atp1a2          0         0        4        3         0</span></span>
<span><span class="co">#&gt; Tnc             0         0        0        0         0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 sdimx    sdimy</span></span>
<span><span class="co">#&gt; cell_1276  -13372.899 15776.37</span></span>
<span><span class="co">#&gt; cell_8890    8941.101 15866.37</span></span>
<span><span class="co">#&gt; cell_691   -14882.899 15896.37</span></span>
<span><span class="co">#&gt; cell_396   -15492.899 15835.37</span></span>
<span><span class="co">#&gt; cell_9818   11308.101 15846.37</span></span>
<span><span class="co">#&gt; cell_11310  14894.101 15810.37</span></span></code></pre></div>
<p>Initialize a SpatialExperiment object and perform basic quality
control. We keep cells with total transcript count within the 5th and
98th percentile:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment.html" class="external-link">SpatialExperiment</a></span><span class="op">(</span>assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">gcm</span><span class="op">)</span>, spatialCoords <span class="op">=</span> <span class="va">locs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># QC based on total counts</span></span>
<span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">thres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">qcstats</span><span class="op">$</span><span class="va">total</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.98</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">qcstats</span><span class="op">$</span><span class="va">total</span> <span class="op">&gt;</span> <span class="va">thres</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">qcstats</span><span class="op">$</span><span class="va">total</span> <span class="op">&lt;</span> <span class="va">thres</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>, <span class="va">keep</span><span class="op">]</span></span></code></pre></div>
<p>Next, perform normalization of the data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Normalization to mean library size</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html" class="external-link">computeLibraryFactors</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">aname</span> <span class="op">&lt;-</span> <span class="st">"normcounts"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="va">aname</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/normalizeCounts.html" class="external-link">normalizeCounts</a></span><span class="op">(</span><span class="va">se</span>, log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="parameters">Parameters<a class="anchor" aria-label="anchor" href="#parameters"></a>
</h2>
<p><em>BANKSY</em> has a few key parameters. We describe these
below.</p>
<div class="section level3">
<h3 id="agf-usage">AGF usage<a class="anchor" aria-label="anchor" href="#agf-usage"></a>
</h3>
<p>For characterising neighborhoods, <em>BANKSY</em> computes the
weighted neighborhood mean (<code>H_0</code>) and the azimuthal Gabor
filter (<code>H_1</code>), which estimates gene expression gradients.
Setting <code>compute_agf=TRUE</code> computes both <code>H_0</code> and
<code>H_1</code>.</p>
</div>
<div class="section level3">
<h3 id="k-geometric">k-geometric<a class="anchor" aria-label="anchor" href="#k-geometric"></a>
</h3>
<p><code>k_geom</code> specifies the number of neighbors used to compute
each <code>H_m</code> for <code>m=0,1</code>. If a single value is
specified, the same <code>k_geom</code> will be used for each feature
matrix. Alternatively, multiple values of <code>k_geom</code> can be
provided for each feature matrix. Here, we use <code>k_geom[1]=15</code>
and <code>k_geom[2]=30</code> for <code>H_0</code> and <code>H_1</code>
respectively. More neighbors are used to compute gradients.</p>
<blockquote>
<p>For datasets generated using Visium v1/v2, use <code>k_geom=18</code>
(or <code>k_geom &lt;- c(18, 18)</code> if
<code>compute_agf = TRUE</code>), since that corresponds to taking as
neighbourhood two concentric rings of spots around each spot.</p>
</blockquote>
<p>We compute the neighborhood feature matrices using normalized
expression (<code>normcounts</code> in the <code>se</code> object).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k_geom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeBanksy.html">computeBanksy</a></span><span class="op">(</span><span class="va">se</span>, assay_name <span class="op">=</span> <span class="va">aname</span>, compute_agf <span class="op">=</span> <span class="cn">TRUE</span>, k_geom <span class="op">=</span> <span class="va">k_geom</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing neighbors...</span></span>
<span><span class="co">#&gt; Spatial mode is kNN_median</span></span>
<span><span class="co">#&gt; Parameters: k_geom=15</span></span>
<span><span class="co">#&gt; Done</span></span>
<span><span class="co">#&gt; Computing neighbors...</span></span>
<span><span class="co">#&gt; Spatial mode is kNN_median</span></span>
<span><span class="co">#&gt; Parameters: k_geom=30</span></span>
<span><span class="co">#&gt; Done</span></span>
<span><span class="co">#&gt; Computing harmonic m = 0</span></span>
<span><span class="co">#&gt; Using 15 neighbors</span></span>
<span><span class="co">#&gt; Done</span></span>
<span><span class="co">#&gt; Computing harmonic m = 1</span></span>
<span><span class="co">#&gt; Using 30 neighbors</span></span>
<span><span class="co">#&gt; Centering</span></span>
<span><span class="co">#&gt; Done</span></span></code></pre></div>
<p><code>computeBanksy</code> populates the <code>assays</code> slot
with <code>H_0</code> and <code>H_1</code> in this instance:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 120 10205 </span></span>
<span><span class="co">#&gt; metadata(1): BANKSY_params</span></span>
<span><span class="co">#&gt; assays(4): counts normcounts H0 H1</span></span>
<span><span class="co">#&gt; rownames(120): Sparcl1 Slc1a2 ... Notch3 Egfr</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(10205): cell_1276 cell_691 ... cell_11635 cell_10849</span></span>
<span><span class="co">#&gt; colData names(4): sample_id sdimx sdimy sizeFactor</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : sdimx sdimy</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="lambda">lambda<a class="anchor" aria-label="anchor" href="#lambda"></a>
</h3>
<p>The <code>lambda</code> parameter is a mixing parameter in
<code>[0,1]</code> which determines how much spatial information is
incorporated for downstream analysis. With smaller values of
<code>lambda</code>, BANKY operates in <em>cell-typing</em> mode, while
at higher levels of <code>lambda</code>, BANKSY operates in
<em>domain-finding</em> mode. As a starting point, we recommend
<code>lambda=0.2</code> for cell-typing and <code>lambda=0.8</code> for
zone-finding, <strong>except for datasets generated using the Visium
v1/v2 technology</strong>, for which we <a href="https://prabhakarlab.github.io/Banksy/articles/multi-sample.html">recommend</a>
<code>lambda=0.2</code> for domain finding. See the note in the tutorial
on the <a href="https://github.com/prabhakarlab/Banksy" class="external-link">main page</a>
for more info.</p>
<p>Here, we run <code>lambda=0</code> which corresponds to non-spatial
clustering, and <code>lambda=0.2</code> for spatially-informed
cell-typing. We compute PCs with and without the AGF
(<code>H_1</code>).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lambda</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runBanksyPCA.html">runBanksyPCA</a></span><span class="op">(</span><span class="va">se</span>, use_agf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>, lambda <span class="op">=</span> <span class="va">lambda</span>, seed <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span></code></pre></div>
<p><code>runBanksyPCA</code> populates the <code>reducedDims</code>
slot, with each combination of <code>use_agf</code> and
<code>lambda</code> provided.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "PCA_M0_lam0"   "PCA_M0_lam0.2" "PCA_M1_lam0"   "PCA_M1_lam0.2"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="clustering-parameters">Clustering parameters<a class="anchor" aria-label="anchor" href="#clustering-parameters"></a>
</h3>
<p>Next, we cluster the BANKSY embedding with Leiden graph-based
clustering. This admits two parameters: <code>k_neighbors</code> and
<code>resolution</code>. <code>k_neighbors</code> determines the number
of k nearest neighbors used to construct the shared nearest neighbors
graph. Leiden clustering is then performed on the resultant graph with
resolution <code>resolution</code>. For reproducibiltiy we set a seed
for each parameter combination.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterBanksy.html">clusterBanksy</a></span><span class="op">(</span><span class="va">se</span>, use_agf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>, lambda <span class="op">=</span> <span class="va">lambda</span>, k_neighbors <span class="op">=</span> <span class="va">k</span>, resolution <span class="op">=</span> <span class="va">res</span>, seed <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span>
<span><span class="co">#&gt; Using seed=1000</span></span></code></pre></div>
<p><code>clusterBanksy</code> populates <code>colData(se)</code> with
cluster labels:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "sample_id"                "sdimx"                   </span></span>
<span><span class="co">#&gt; [3] "sdimy"                    "sizeFactor"              </span></span>
<span><span class="co">#&gt; [5] "clust_M0_lam0_k50_res1"   "clust_M0_lam0.2_k50_res1"</span></span>
<span><span class="co">#&gt; [7] "clust_M1_lam0_k50_res1"   "clust_M1_lam0.2_k50_res1"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="comparing-cluster-results">Comparing cluster results<a class="anchor" aria-label="anchor" href="#comparing-cluster-results"></a>
</h2>
<p>To compare clustering runs visually, different runs can be relabeled
to minimise their differences with <code>connectClusters</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/connectClusters.html">connectClusters</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; clust_M1_lam0_k50_res1 --&gt; clust_M0_lam0_k50_res1</span></span>
<span><span class="co">#&gt; clust_M0_lam0.2_k50_res1 --&gt; clust_M1_lam0_k50_res1</span></span>
<span><span class="co">#&gt; clust_M1_lam0.2_k50_res1 --&gt; clust_M0_lam0.2_k50_res1</span></span></code></pre></div>
<p>Visualise spatial coordinates with cluster labels.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cnames</span> <span class="op">&lt;-</span> <span class="va">cnames</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^clust"</span>, <span class="va">cnames</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cnames</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cnm</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">se</span>, x <span class="op">=</span> <span class="st">"sdimx"</span>, y <span class="op">=</span> <span class="st">"sdimy"</span>, point_size <span class="op">=</span> <span class="fl">0.1</span>, colour_by <span class="op">=</span> <span class="va">cnm</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">cnm</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">cplots</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/parameter-selection-spatial-1.png" width="90%"></p>
<p>Compare all cluster outputs with <code>compareClusters</code>. This
function computes pairwise cluster comparison metrics between the
clusters in <code>colData(se)</code> based on adjusted Rand index
(ARI):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compareClusters.html">compareClusters</a></span><span class="op">(</span><span class="va">se</span>, func <span class="op">=</span> <span class="st">"ARI"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                          clust_M0_lam0_k50_res1 clust_M0_lam0.2_k50_res1</span></span>
<span><span class="co">#&gt; clust_M0_lam0_k50_res1                    1.000                     0.67</span></span>
<span><span class="co">#&gt; clust_M0_lam0.2_k50_res1                  0.670                     1.00</span></span>
<span><span class="co">#&gt; clust_M1_lam0_k50_res1                    1.000                     0.67</span></span>
<span><span class="co">#&gt; clust_M1_lam0.2_k50_res1                  0.747                     0.87</span></span>
<span><span class="co">#&gt;                          clust_M1_lam0_k50_res1 clust_M1_lam0.2_k50_res1</span></span>
<span><span class="co">#&gt; clust_M0_lam0_k50_res1                    1.000                    0.747</span></span>
<span><span class="co">#&gt; clust_M0_lam0.2_k50_res1                  0.670                    0.870</span></span>
<span><span class="co">#&gt; clust_M1_lam0_k50_res1                    1.000                    0.747</span></span>
<span><span class="co">#&gt; clust_M1_lam0.2_k50_res1                  0.747                    1.000</span></span></code></pre></div>
<p>or normalized mutual information (NMI):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compareClusters.html">compareClusters</a></span><span class="op">(</span><span class="va">se</span>, func <span class="op">=</span> <span class="st">"NMI"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                          clust_M0_lam0_k50_res1 clust_M0_lam0.2_k50_res1</span></span>
<span><span class="co">#&gt; clust_M0_lam0_k50_res1                    1.000                    0.741</span></span>
<span><span class="co">#&gt; clust_M0_lam0.2_k50_res1                  0.741                    1.000</span></span>
<span><span class="co">#&gt; clust_M1_lam0_k50_res1                    1.000                    0.741</span></span>
<span><span class="co">#&gt; clust_M1_lam0.2_k50_res1                  0.782                    0.915</span></span>
<span><span class="co">#&gt;                          clust_M1_lam0_k50_res1 clust_M1_lam0.2_k50_res1</span></span>
<span><span class="co">#&gt; clust_M0_lam0_k50_res1                    1.000                    0.782</span></span>
<span><span class="co">#&gt; clust_M0_lam0.2_k50_res1                  0.741                    0.915</span></span>
<span><span class="co">#&gt; clust_M1_lam0_k50_res1                    1.000                    0.782</span></span>
<span><span class="co">#&gt; clust_M1_lam0.2_k50_res1                  0.782                    1.000</span></span></code></pre></div>
<p>See <code><a href="../reference/compareClusters.html">?compareClusters</a></code> for the full list of comparison
measures.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>Vignette runtime:</p>
<pre><code><span><span class="co">#&gt; Time difference of 27.6314 secs</span></span></code></pre>
<details><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.2.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Detroit</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] cowplot_1.1.3               scater_1.30.1              </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.4.4               scuttle_1.12.0             </span></span>
<span><span class="co">#&gt;  [5] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.54.1        GenomeInfoDb_1.38.6        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.36.0              S4Vectors_0.40.2           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.2.0           Banksy_0.99.12             </span></span>
<span><span class="co">#&gt; [17] BiocStyle_2.30.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] bitops_1.0-7              gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [3] rlang_1.1.3               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;  [5] compiler_4.3.2            sccore_1.0.4             </span></span>
<span><span class="co">#&gt;  [7] DelayedMatrixStats_1.24.0 systemfonts_1.0.5        </span></span>
<span><span class="co">#&gt;  [9] vctrs_0.6.5               stringr_1.5.1            </span></span>
<span><span class="co">#&gt; [11] pkgconfig_2.0.3           crayon_1.5.2             </span></span>
<span><span class="co">#&gt; [13] fastmap_1.1.1             magick_2.8.2             </span></span>
<span><span class="co">#&gt; [15] XVector_0.42.0            labeling_0.4.3           </span></span>
<span><span class="co">#&gt; [17] utf8_1.2.4                rmarkdown_2.25           </span></span>
<span><span class="co">#&gt; [19] ggbeeswarm_0.7.2          ragg_1.2.7               </span></span>
<span><span class="co">#&gt; [21] purrr_1.0.2               xfun_0.42                </span></span>
<span><span class="co">#&gt; [23] zlibbioc_1.48.0           cachem_1.0.8             </span></span>
<span><span class="co">#&gt; [25] beachmat_2.18.0           jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt; [27] highr_0.10                DelayedArray_0.28.0      </span></span>
<span><span class="co">#&gt; [29] BiocParallel_1.36.0       irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt; [31] parallel_4.3.2            aricode_1.0.3            </span></span>
<span><span class="co">#&gt; [33] R6_2.5.1                  bslib_0.6.1              </span></span>
<span><span class="co">#&gt; [35] stringi_1.8.3             leidenAlg_1.1.2          </span></span>
<span><span class="co">#&gt; [37] jquerylib_0.1.4           Rcpp_1.0.12              </span></span>
<span><span class="co">#&gt; [39] bookdown_0.37             knitr_1.45               </span></span>
<span><span class="co">#&gt; [41] Matrix_1.6-5              igraph_2.0.1.1           </span></span>
<span><span class="co">#&gt; [43] tidyselect_1.2.0          viridis_0.6.5            </span></span>
<span><span class="co">#&gt; [45] rstudioapi_0.15.0         abind_1.4-5              </span></span>
<span><span class="co">#&gt; [47] yaml_2.3.8                codetools_0.2-19         </span></span>
<span><span class="co">#&gt; [49] lattice_0.22-5            tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [51] withr_3.0.0               evaluate_0.23            </span></span>
<span><span class="co">#&gt; [53] desc_1.4.3                mclust_6.0.1             </span></span>
<span><span class="co">#&gt; [55] pillar_1.9.0              BiocManager_1.30.22      </span></span>
<span><span class="co">#&gt; [57] generics_0.1.3            dbscan_1.1-12            </span></span>
<span><span class="co">#&gt; [59] RCurl_1.98-1.14           sparseMatrixStats_1.14.0 </span></span>
<span><span class="co">#&gt; [61] munsell_0.5.0             scales_1.3.0             </span></span>
<span><span class="co">#&gt; [63] glue_1.7.0                tools_4.3.2              </span></span>
<span><span class="co">#&gt; [65] BiocNeighbors_1.20.2      data.table_1.15.0        </span></span>
<span><span class="co">#&gt; [67] ScaledMatrix_1.10.0       fs_1.6.3                 </span></span>
<span><span class="co">#&gt; [69] grid_4.3.2                colorspace_2.1-0         </span></span>
<span><span class="co">#&gt; [71] GenomeInfoDbData_1.2.11   RcppHungarian_0.3        </span></span>
<span><span class="co">#&gt; [73] beeswarm_0.4.0            BiocSingular_1.18.0      </span></span>
<span><span class="co">#&gt; [75] vipor_0.4.7               cli_3.6.2                </span></span>
<span><span class="co">#&gt; [77] rsvd_1.0.5                textshaping_0.3.7        </span></span>
<span><span class="co">#&gt; [79] fansi_1.0.6               viridisLite_0.4.2        </span></span>
<span><span class="co">#&gt; [81] S4Arrays_1.2.0            dplyr_1.1.4              </span></span>
<span><span class="co">#&gt; [83] uwot_0.1.16               gtable_0.3.4             </span></span>
<span><span class="co">#&gt; [85] sass_0.4.8                digest_0.6.34            </span></span>
<span><span class="co">#&gt; [87] ggrepel_0.9.5             SparseArray_1.2.4        </span></span>
<span><span class="co">#&gt; [89] farver_2.1.1              rjson_0.2.21             </span></span>
<span><span class="co">#&gt; [91] memoise_2.0.1             htmltools_0.5.7          </span></span>
<span><span class="co">#&gt; [93] pkgdown_2.0.7             lifecycle_1.0.4</span></span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vipul Singhal, Joseph Lee.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
